Python/compile.c

```diff
@@ -5421,6 +5452,601 @@ compiler_slice(struct compiler *c, expr_ty s)
     return 1;
 }
 
+
+// PEP 634: Structural Pattern Matching
+
+// To keep things simple, all compiler_pattern_* routines follow the convention
+// of preserving TOS (the subject for the given pattern) and pushing either True
+// (match) or False (no match) on top of it. We do this even for irrefutable
+// patterns; the idea is that it's much easier to smooth out any redundant
+// pushing, popping, and jumping in the peephole optimizer than to detect or
+// predict it here.
+
+
+#define WILDCARD_CHECK(N) \
+    ((N)->kind == Name_kind && \
+    _PyUnicode_EqualToASCIIString((N)->v.Name.id, "_"))
+
+
+static int
+pattern_helper_load_attr(struct compiler *c, expr_ty p, pattern_context *pc)
+{
+    assert(p->kind == Attribute_kind);
+    assert(p->v.Attribute.ctx == Load);
+    if (p->v.Attribute.value->kind == Attribute_kind) {
+        RETURN_IF_FALSE(pattern_helper_load_attr(c, p->v.Attribute.value, pc));
+    }
+    else {
+        assert(p->v.Attribute.value->kind == Name_kind);
+        assert(p->v.Attribute.value->v.Name.ctx == Load);
+        RETURN_IF_FALSE(compiler_nameop(c, p->v.Attribute.value->v.Name.id, Load));
+    }
+    ADDOP_NAME(c, LOAD_ATTR, p->v.Attribute.attr, names);
+    return 1;
+}
+
+
+static int
+pattern_helper_load_constant(struct compiler *c, expr_ty p, pattern_context *pc)
+{
+    assert(p->kind == Constant_kind);
+    assert(p->v.Constant.value == Py_None ||
+           PyBool_Check(p->v.Constant.value) ||
+           PyBytes_CheckExact(p->v.Constant.value) ||
+           PyComplex_CheckExact(p->v.Constant.value) ||
+           PyFloat_CheckExact(p->v.Constant.value) ||
+           PyLong_CheckExact(p->v.Constant.value) ||
+           PyUnicode_CheckExact(p->v.Constant.value));
+    ADDOP_LOAD_CONST(c, p->v.Constant.value);
+    return 1;
+}
+
+
+static int
+pattern_helper_store_name(struct compiler *c, identifier n, pattern_context *pc)
+{
+    assert(!_PyUnicode_EqualToASCIIString(n, "_"));
+    // Can't assign to the same name twice:
+    if (!pc->stores) {
+        RETURN_IF_FALSE(pc->stores = PySet_New(NULL));
+    }
+    else {
+        int duplicate = PySet_Contains(pc->stores, n);
+        if (duplicate < 0) {
+            return 0;
+        }
+        if (duplicate) {
+            const char *e = "multiple assignments to name %R in pattern";
+            return compiler_error(c, e, n);
+        }
+    }
+    RETURN_IF_FALSE(!PySet_Add(pc->stores, n));
+    RETURN_IF_FALSE(compiler_nameop(c, n, Store));
+    return 1;
+}
+
+
+static int
+compiler_pattern_subpattern(struct compiler *c, expr_ty p, pattern_context *pc)
+{
+    // Like compiler_pattern, but turn off checks for irrefutability.
+    int allow_irrefutable = pc->allow_irrefutable;
```

Comment:
This is more an observation than a request:
The test for irrefutability (is that a real word?) seem to be passed down the stack.
Would passing it as an argument make sense?

Reply:
Everything in `pattern_context` originally started as explicit arguments. Eventually I got tired of keeping track of everything, and just threw all of that state into the struct.

Personally, I find it easier to reason about all of these routines if they have the same signature, stack effect, etc. So I'm slightly in favor of keeping it this way... especially since I anticipate that `pattern_context` will grow more members in the future (to facilitate more efficient matching strategies, provide better warnings/errors, etc.).